crates/curp/src/client/stream.rs

```diff
@@ -0,0 +1,113 @@
+use std::{sync::Arc, time::Duration};
+
+use futures::Future;
+use tracing::{debug, warn};
+
+use crate::rpc::{connect::ConnectApi, CurpError, Redirect};
+
+use super::state::State;
+
+/// Stream client config
+#[derive(Debug)]
+pub(super) struct StreamingConfig {
+    /// Heartbeat interval
+    heartbeat_interval: Duration,
+}
+
+impl StreamingConfig {
+    /// Create a stream client config
+    pub(super) fn new(heartbeat_interval: Duration) -> Self {
+        Self { heartbeat_interval }
+    }
+}
+
+/// Stream client
+#[derive(Debug)]
+pub(super) struct Streaming {
+    /// Shared client state
+    pub(super) state: Arc<State>,
+    /// Stream client config
+    config: StreamingConfig,
+}
+
+impl Streaming {
+    /// Create a stream client
+    pub(super) fn new(state: Arc<State>, config: StreamingConfig) -> Self {
+        Self { state, config }
+    }
+
+    /// Take an async function and map to the remote leader, hang up when no leader found or
+    /// the leader is itself.
+    async fn map_remote_leader<R, F: Future<Output = Result<R, CurpError>>>(
+        &self,
+        f: impl FnOnce(Arc<dyn ConnectApi>) -> F,
+    ) -> Result<R, CurpError> {
+        loop {
+            let Some(leader_id) = self.state.leader_id().await else {
+                debug!("cannot find the leader id in state, wait for leadership update");
+                self.state.leader_notifier().listen().await;
+                continue;
+            };
+            if let Some(local_id) = self.state.local_server_id() {
+                if leader_id == local_id {
+                    self.state.check_gen_local_client_id().await;
+                    debug!("skip keep heartbeat for local connection, wait for leadership update");
+                    self.state.leader_notifier().listen().await;
+                    continue;
+                }
+            }
+            return self.state.map_server(leader_id, f).await;
+        }
+    }
+
+    /// Keep heartbeat
+    pub(super) async fn keep_heartbeat(&self) {
+        /// Prevent lock contention when leader crashed or some unknown errors
+        const RETRY_DELAY: Duration = Duration::from_millis(100);
+
+        loop {
+            // is heartbeat task cancellation safety?
+            let heartbeat = self.map_remote_leader::<(), _>(|conn| async move {
```

Comment:
I think heartbeat is cancel safe as it doesn't depend on external states.

Reply:
comment removed